aws/table_aws_ecs_container_instance.go

```diff
@@ -0,0 +1,279 @@
+				Hydrate:     getEcsContainerInstance,
+				Transform:   transform.From(containerInstanceTurbotTags),
+			},
+			{
+				Name:        "title",
+				Description: resourceInterfaceDescription("title"),
+				Type:        proto.ColumnType_STRING,
+				Hydrate:     getEcsContainerInstance,
+				Transform:   transform.FromField("ContainerInstanceArn"),
+			},
+			{
+				Name:        "akas",
+				Description: resourceInterfaceDescription("akas"),
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     getEcsContainerInstance,
+				Transform:   transform.FromField("ContainerInstanceArn").Transform(arnToAkas),
+			},
+		}),
+	}
+}
+
+//// LIST FUNCTION
+
+func listEcsContainerInstances(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("listEcsContainerInstances")
+
+	var region string
+	matrixRegion := plugin.GetMatrixItem(ctx)[matrixKeyRegion]
+	if matrixRegion != nil {
+		region = matrixRegion.(string)
+	}
+
+	// Create Session
+	svc, err := EcsService(ctx, d, region)
+	if err != nil {
+		return nil, err
+	}
+
+	// EcsContainerInstance is a sub resource of an EcsCluster, we need the cluster ARN to list these.
+	var clusterArn string
+	if h.Item != nil {
+		clusterArn = *h.Item.(*ecs.Cluster).ClusterArn
+	} else {
+		clusterArn = d.KeyColumnQuals["cluster_arn"].GetStringValue()
+	}
+
+	// execute list call
+	err = svc.ListContainerInstancesPages(
+		&ecs.ListContainerInstancesInput{
+			Cluster: aws.String(clusterArn),
+		},
+		func(page *ecs.ListContainerInstancesOutput, isLast bool) bool {
+			for _, arn := range page.ContainerInstanceArns {
+				d.StreamListItem(ctx, &ecs.ContainerInstance{
+					ContainerInstanceArn: arn,
+				})
+			}
+			return !isLast
+		},
+	)
+	return nil, err
+}
+
+//// HYDRATE FUNCTIONS
+
+func getEcsContainerInstance(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("describeEcsContainerInstance")
+	var region string
+	matrixRegion := plugin.GetMatrixItem(ctx)[matrixKeyRegion]
+	if matrixRegion != nil {
+		region = matrixRegion.(string)
+	}
+
+	var containerInstanceArn string
+	if h.Item != nil {
+		containerInstanceArn = *h.Item.(*ecs.ContainerInstance).ContainerInstanceArn
+	} else {
+		containerInstanceArn = d.KeyColumnQuals["container_instances_arn"].GetStringValue()
+	}
+
+	// This isn't returned from the ListContainerInstances API and is needed to run DescribeContainerInstancesInput.
+	// We can however construct it manually from the containerInstanceArn
+	clusterArn := clusterArnFromContainerInstanceArn(containerInstanceArn)
+
+	// Create Session
+	svc, err := EcsService(ctx, d, region)
+	if err != nil {
+		return nil, err
+	}
+
+	// execute list call
+	input := &ecs.DescribeContainerInstancesInput{
+		Cluster: aws.String(clusterArn),
+		ContainerInstances: []*string{aws.String(containerInstanceArn)},
+		Include: []*string{aws.String("TAGS")},
+	}
```

Comment:
DescribeContainerInstances can take up to 100 ContainerInstanceArns and returns the results from all of them. Is there a way to take advantage of this to reduce calls?

Reply:
Interesting case... most apis take one at a time so we’re broadly optimized for that case.

when you do the list call and start streaming back the items, steampipe starts doing the get calls immediately in parallel. It’s also smart about only doing hydrate calls of the column has been requested.

You could sacrifice that stuff, and do more work in the list query? Collect the arns looping through the list and then do a big get call. Then stream the results of the multi-get call back. That would do two api call sequentially for the list action, but would avoid get per row.

Reply:
Yeah I think that makes sense, will go ahead and do it this way.